src/tablenavigation.js

```diff
@@ -0,0 +1,480 @@
+
+			if ( nextRect && rect.contains( nextRect ) ) {
+				continue;
+			}
+
+			if ( boundaryVerticalPosition === undefined ) {
+				boundaryVerticalPosition = Math.round( isForward ? rect.top : rect.bottom );
+				continue;
+			}
+
+			// Let's check if this rect is in new line.
+			if ( isForward ) {
+				if ( Math.round( rect.bottom ) <= boundaryVerticalPosition ) {
+					return false;
+				}
+
+				boundaryVerticalPosition = Math.min( boundaryVerticalPosition, rect.top );
+			} else {
+				if ( Math.round( rect.top ) >= boundaryVerticalPosition ) {
+					return false;
+				}
+
+				boundaryVerticalPosition = Math.max( boundaryVerticalPosition, rect.bottom );
+			}
+		}
+
+		return true;
+	}
+
+	/**
+	 * Moves the selection from the given table cell in the specified direction.
+	 *
+	 * @private
+	 * @param {module:engine/model/element~Element} tableCell The table cell to start the selection navigation.
+	 * @param {'left'|'up'|'right'|'down'} direction Direction in which selection should move.
+	 */
+	_navigateFromCellInDirection( tableCell, direction ) {
+		const model = this.editor.model;
+
+		const table = findAncestor( 'table', tableCell );
+		const tableMap = [ ...new TableWalker( table, { includeSpanned: true } ) ];
+		const { row: lastRow, column: lastColumn } = tableMap[ tableMap.length - 1 ];
+
+		const currentCellInfo = tableMap.find( ( { cell } ) => cell == tableCell );
+		let { row, column } = currentCellInfo;
+
+		switch ( direction ) {
+			case 'left':
+				column--;
+				break;
+
+			case 'up':
+				row--;
+				break;
+
+			case 'right':
+				column += currentCellInfo.colspan || 1;
+				break;
+
+			case 'down':
+				row += currentCellInfo.rowspan || 1;
+				break;
+		}
+
+		const isOutsideVertically = row < 0 || row > lastRow;
+		const isBeforeFirstCell = column < 0 && row <= 0;
+		const isAfterLastCell = column > lastColumn && row >= lastRow;
+
+		// Note that if the last table cell is row-spanned then isAfterLastCell will never be true but we don't know
+		// if user was navigating on the last row or not, so let's allow him to stay in the table.
+
+		if ( isOutsideVertically || isBeforeFirstCell || isAfterLastCell ) {
+			model.change( writer => {
+				writer.setSelection( writer.createRangeOn( table ) );
+			} );
+
+			return;
+		}
+
+		if ( column < 0 ) {
+			column = lastColumn;
+			row--;
+		} else if ( column > lastColumn ) {
+			column = 0;
+			row++;
+		}
+
+		const cellToSelect = tableMap.find( cellInfo => cellInfo.row == row && cellInfo.column == column ).cell;
+		const isForward = [ 'right', 'down' ].includes( direction );
+		const positionToSelect = model.createPositionAt( cellToSelect, isForward ? 0 : 'end' );
+
+		model.change( writer => {
+			writer.setSelection( positionToSelect );
+		} );
+	}
+}
+
+// Returns 'true' if provided key code represents one of the arrow keys.
+//
+// @private
+// @param {Number} keyCode
+// @returns {Boolean}
+function isArrowKeyCode( keyCode ) {
```

Comment:
I think it's time we had an utils for it in `keyboard.js`. Use cases:

```
6 results - 3 files

⚙️ engine • src/view/observer/fakeselectionobserver.js:
   52: 			if ( selection.isFake && _isArrowKeyCode( data.keyCode ) && this.isEnabled ) {
  119: function _isArrowKeyCode( keyCode ) {

table • src/tablenavigation.js:
  165: 		if ( isArrowKeyCode( keyCode ) ) {
  460: function isArrowKeyCode( keyCode ) {

widget • src/widget.js:
  173: 		if ( isArrowKeyCode( keyCode ) ) {
  421: function isArrowKeyCode( keyCode ) {

```

(this could be a follow-up).

Reply:
Created a follow-up https://github.com/ckeditor/ckeditor5/issues/6648.